### DISCUSSION TITLE
Empty ChangeSet in onFlush event when setting OneToOne relation.

### DISCUSSION THREAD

**Discussion by pawelsmolenski** [mikro-orm/mikro-orm#3732] [Q&A]
When updating an entity, setting a non-null value of the OneToOne relation causes the changesets inside `onFlush` event handler to be an empty array, while at the same time changeset of `beforeUpdate` and `afterUpdate` events contains information about that modified value. Not sure if this is an expected behavior – especially that doing the same, but clearing that relation (i.e. setting null value on its property) gives "proper" changesets not only in `beforeUpdate` and `afterUpdate` but all of these handlers.

**To Reproduce**
Not sure how to include a working code, but will try anyway with those fragments:

Two entities (`Display` and `Device`) with one to one, nullable relationship between them:
```typescript
@Entity
export class Display extends Base {
    @OneToOne({ 
        entity: () => Device, 
        inversedBy: "display", 
        nullable: true,
        owner: true, 
    })
    device?: Device;

    ...
}

@Entity
export class Device extends Base {
    @OneToOne({
        entity: () => Display,
        mappedBy: "device",
        nullable: true,
        owner: false,
    })
    display?: Display;

    ...
}
```

Entity lifecycle subscriber, "listening" on global `onFlush` and `Display` specific (`beforeUpdate` and `afterUpdate`) events:
```typescript
@Injectable
export class DisplaySubscriber implements EventSubscriber<Display> {
    constructor(
        @Inject(MIKRO_ORM_CONTENT)
        private readonly orm: MikroORM<PostgreSqlDriver>,
    ) {
        this.orm.em.getEventManager().registerSubscriber(this);
    }

    public getSubscribedEntities(): Array<EntityName<Display>> {
        return [ Display ];
    }

    public async afterUpdate(args: EventArgs<Display>): Promise<void> {
        this.logger.debug("afterUpdate...", args.changeSet);
    }

    public async beforeUpdate(args: EventArgs<Display>): Promise<void> {
        this.logger.debug("beforeUpdate...", args.changeSet);
    }

    public async onFlush(args: FlushEventArgs): Promise<void> {
        this.logger.debug("onFlush...", args.uow.getChangeSets());
    }

    ...
}
```

In my opinion the problem occurs while updating `Display` with a new or modified relation to a `Device` entity:
```typescript
const display: Display = await em.getRepository(Display).findOneOrFail({ id: "6619a86b-a675-401a-8729-5d6c2331a597" });
display.device = em.getRepository(Device).getReference("a2fb4afa-895d-47b1-8954-26d9657f706f");

await em.flush();
```
...when changesets of `onFlush` event handler is an empty array, as shown below:

```
onFlush...      changeSets: (0) []
beforeUpdate... changeSet:  { payload: { device: "a2fb4afa-895d-47b1-8954-26d9657f706f" }, ... }
afterUpdate...  changeSet:  { payload: { device: "a2fb4afa-895d-47b1-8954-26d9657f706f" }, ... }
```
SQL query underneath:
```sql
update "displays" set "device_id" = 'a2fb4afa-895d-47b1-8954-26d9657f706f' where "id" = '6619a86b-a675-401a-8729-5d6c2331a597'
```

I have been struggling with this problem, trying to simplify the code to eliminate influence of any external dependencies, eventually giving up. I even was close to believing that this behavior is as designed, but fact that setting null value of the same relation (`Display.device`):
```typescript
const display: DisplayEntity = await em.getRepository(Display).findOneOrFail({ id: input.id });
display.device = null;

await em.flush();
```
...gives onFlush changesets which contains expected data:
```
onFlush...      changeSets: (1) [ { payload: { device: undefined }, ... } ]
beforeUpdate... changeSet:  { payload: { device: undefined }, ... }
afterUpdate...  changeSet:  { payload: { device: undefined }, ... }
```
SQL query:
```sql
update "displays" set "device_id" = NULL where "id" = '6619a86b-a675-401a-8729-5d6c2331a597'
```
...gives me a hope this is a bug.

**Expected behavior**
Changeset(s) of `onFlush` event handler should contain newly set values of OneToOne relation, despite they are null or a reference to an entity.

**Additional context**
Problem does not affect ManyToOne relationships. I.e. changing relation to the `ManyToOne` makes changeset appear in `onFlush` hook arguments.

**Versions**

| Dependency | Version |
| - | - |
| node | 16.18.0 |
| typescript | 4.8.2 |
| mikro-orm | 5.3.1, 5.5.3 |
| your-driver | PostgreSQL |

**Comment by pawelsmolenski** [comment]
More details after some additional investigation:

I modified entities and added a ManyToOne relationship alongside OneToOne for test purposes:

```typescript
@Entity
export class Display extends Base {
    @OneToOne({ 
        entity: () => Device, 
        inversedBy: "display", 
        nullable: true,
        owner: true, 
    })
    device?: Device;
    
    @ManyToOne({ 
        entity: () => Display, 
        onDelete: "set null", 
        nullable: true 
    })
    cloneFrom?: Display;

    ...
}

@Entity
export class Device extends Base {
    @OneToOne({
        entity: () => Display,
        mappedBy: "device",
        nullable: true,
        owner: false,
    })
    display?: Display;

    ...
}
```

When simultaneously updating both of `Display` relationships (`Display.device` (1:1) and `Display.cloneFrom` (m:1)) with new FK values:

```typescript
const display: Display = await em.getRepository(Display).findOneOrFail({ id: "6619a86b-a675-401a-8729-5d6c2331a597" });

display.device = em.getRepository(Device).getReference("a2fb4afa-895d-47b1-8954-26d9657f706f");
display.cloneFrom = em.getRepository(Display).getReference("bb8aa2d0-56bf-4bfb-b72f-722c63a1ea7f");

await em.flush();
```

I am getting two update queries:
```
DEBUG [ORM-Content] [query] begin
DEBUG [ORM-Content] [query] update "displays" set "clone_from_id" = 'bb8aa2d0-56bf-4bfb-b72f-722c63a1ea7f' where "id" = '6619a86b-a675-401a-8729-5d6c2331a597' [took 158 ms]
DEBUG [ORM-Content] [query] update "displays" set "device_id" = 'a2fb4afa-895d-47b1-8954-26d9657f706f' where "id" = '6619a86b-a675-401a-8729-5d6c2331a597' [took 25 ms]
DEBUG [ORM-Content] [query] commit
```
and only a single property in payload of `onFlush` changesets:
```js
[ 
    { 
        payload: { 
            cloneFrom: 'bb8aa2d0-56bf-4bfb-b72f-722c63a1ea7f' 
        }
        ...
    }
]
```

But when I try then to set null to both of them:
```typescript
const display: Display = await em.getRepository(Display).findOneOrFail({ id: "6619a86b-a675-401a-8729-5d6c2331a597" });

display.device = null;
display.cloneFrom = null;

await em.flush();
```

I am getting a strange payload (notice `undefined` instead of `null`, probably):
```js
[ 
    {
        entity: {
            id: "6619a86b-a675-401a-8729-5d6c2331a597",
            ...
        }, 
        name: 'Display',
        payload: { 
            cloneFrom: null,
            device: undefined
        }
        ...
    }
]
```

and one query this time:
```
DEBUG [ORM-Content] [query] begin
DEBUG [ORM-Content] [query] update "displays" set "clone_from_id" = NULL, "device_id" = NULL where "id" = '6619a86b-a675-401a-8729-5d6c2331a597' [took 268 ms]
DEBUG [ORM-Content] [query] commit
```

P.S. It is worth noticing, that filling both relationships with FK and flushing them at once, what causes separation of this update into two update operations/queries (as described above) also triggers `before/afterUpdate` twice – one for each of those relationships. I am not concerned about this too much, as I am forced to use onFlush hook anyway – to bo able to create additional entities in response to that update, but in other cases in might be an issue, I suppose.

**Comment by B4nan** [comment]
Do I read this right that the queries are produced correctly, you just don't like how the hooks are fired? 

I'd appreciate if you compile this into a single script I can run instead of 10 small disconnected pieces, as in the regression tests:

https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

(it does not have to be a test case, but it should be a complete runnable script, you can as well try to use https://stackblitz.com/edit/mikro-orm-repro)

 > I modified entities and added a ManyToOne relationship alongside OneToOne for test purposes:

That modified entity definition represents two owning sides (so two database column, each in one table), not really comparable to a bidirectional 1:1 relation, which has only one database column.

**Comment by pawelsmolenski** [comment]
@B4nan Thank you for your answer. I tried to write simplified test case (https://stackblitz.com/edit/mikro-orm-repro-5ynnb5), but cannot reproduce same behavior of not including changed reference of OneToOne relationship in `ChangeSet[]` of `onFlush` hook. Also, I cannot achieve the same separation of updates when additionally and simultaneously updating another relationship (ManyToOne in the example). Everything seems to work as expected in this simple example.

Sorry for not finishing the code and using console output for "visual" validation, but not sure how to easily analyze side effects from event hooks. Nevertheless, even at this stage I got a proof that something is wrong with my setup.

I will try to further dismantle my project, until I will find the reason of that behavior. Regards.

**Comment by pawelsmolenski** [comment]
Looked into internals of flushing procedure and for some reason, during update of OneToOne relationship from FK to null value, the "new entity data" (prepared by the ChangeSetComputer Comparator) is lacking that key/property. I suppose it should contain a null value there – as it is in case of ManyToOne relationship. Same happens in an reverse direction: When updating from null to FK, the "original entity data" has `undefined` under OneToOne property instead of a `null` value.

Not sure why it happens (will investigate, maybe schema is wrongly defined), but it explains why when updating from FK to null, the payload of `onFlush` changeset has different values (i.e. `null` and `undefined`) for IMO similar relationships:

```js
[ 
    {
        payload: { 
            cloneFrom: null,
            device: undefined
        }
        ...
    }
]
```

I also noticed that despite that "internal" data inconsistency, payload of changeset is initially calculated properly and for some time it contains both relationships, but information about the new OneToOne relationship is lost somewhere in the later process... TBC

**Comment by pawelsmolenski** [comment]
Ok, it seems that mystery may be solved: 
```
// when changing a unique nullable property (or a 1:1 relation), we can't do it in a single query as it would cause unique constraint violations
```
https://github.com/mikro-orm/mikro-orm/blob/0741633177cee3b437d00e375e1d1c18065ffaa4/packages/core/src/unit-of-work/UnitOfWork.ts#L553

As I understand it, this procedure removes the changed OneToOne FK from the changeset payload and is an explanation of the separated SQL query. It also clarifies why it works "normally" in an opposite direction – when the relationship FK is removed (despite the `null`/`undefined` behavior)... But, tell me if I am wrong, it also causes that information about that change (`null` to `FK`) never reaches the `onFlush` hook.

**Reply by B4nan** [reply]
Yeah, that sounds accurate, extra updates are handled separately. You can get list of scheduled extra updates via `uow.getExtraUpdates()`. Similarly collection updates are not part of the changesets (`uow.getCollectionUpdates()`).

**Reply by pawelsmolenski** [reply]
Thank you again!

Do you predict that one day someone would implement more uniform structure for passing changeset payload in hooks or should it be considered a final form for accessing this data in event handlers?

**Comment by pawelsmolenski** [comment]
I am trying to reverse–engineer "SnapshotGenerator" of `EntityComparator` to understand why unset nullable OneToOne relationship  is represented in changelog payload as `undefined` while similar OneToMany relationship is `null`... I can see that PropertySnapshots are constructed differently:

```js
if (
    typeof entity.cloneFrom !== "undefined" &&
    (entity.cloneFrom == null || entity.cloneFrom.__helper?.hasPrimaryKey())
) {
    ret.cloneFrom = getPrimaryKeyValues_cloneFrom(entity.cloneFrom);
}
```
vs
```js
if (
    typeof entity.device !== "undefined" &&
    entity.device != null &&
    (entity.device == null || entity.device.__helper?.hasPrimaryKey())
) {
    ret.device = getPrimaryKeyValues_device(entity.device);
}
```
...but do not know why. Any hints?

...

It seems that defining reverse side of relationship with `inversedBy` option causes this behavior, but again – why?
If `inversedBy` is not defined then payload of an update from `FK` to `null` looks more legit for me:

```ts
@Entity
export class Display extends Base {
    @OneToOne({ 
        entity: () => Device, 
        nullable: true,
        owner: true, 
    })
    device?: Device;
    
    @ManyToOne({ 
        entity: () => Display, 
        onDelete: "set null", 
        nullable: true 
    })
    cloneFrom?: Display;

    ...
}

@Entity
export class Device extends Base {
    ...
}
```

```js
 [ 
    {
        payload: { 
            cloneFrom: null,
            device: null // Yippee!
        }
        ...
    }
]
```

**Reply by B4nan** [reply]
I don't think the behavior you don't like is a bug. There should be no key in the payload if we schedule an extra update, otherwise, it would still trigger update of that field (but to `null` value), while it should just postpone the update of the unique column to the end of the flush.

Two notes regarding the inversedBy/mappedBy presence:
- Your owning side should have `inversedBy` defined. In this example you don't have any bidirectional 1:1 property, only the owner - hence no `inversedBy` or `mappedBy`.
- While it probably works the same (because you specify `owner` explicitly), your original entity definition contained bidirectional 1:1, both using `mappedBy` - you should use `inversedBy` on the owning side (to point to the inverse), and `mappedBy` on the inverse side (pointing to owner).

**Reply by pawelsmolenski** [reply]
Sorry, if I raised entropy level suggesting this was a bug. I should rather say this is a cognitive problem for me. From the user's point of view, almost the same inputs were giving different outputs without any warnings. Now I understand, that in event hooks we are dealing with a kind of "internal" data, that is not meant to be "human readable". Once again – thank you for your support.

P.S. I fixed that `mappedBy`/`inversedBy` mistake (always had problems with remembering which is what ;)). Thanks for pointing that out and sorry if it affected readability of my question. Luckily for me it was introduced only in discussion (if my boss is reading this).

**Reply by B4nan** [reply]
The extra condition you don't like is added because the property is defined as setter - on the prototype level every relation is defined as get/set pair so we can track changes. 

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/EntityComparator.ts#L315-L319

I don't remember the exact reason, but If I try to remove that extra condition, the test for #1523 breaks.

edit: And no need to be sorry :] Those things are quite complex to understand even for me, as the code is often years old. But that's why I consistently add tests for (almost) every issue/change, so I know something will not break because of a refactoring I am not 100% sure about.

**Reply by pawelsmolenski** [reply]
P.S. Please, do not bother yourself – maybe someone else could explain to me, why my test example (https://stackblitz.com/edit/mikro-orm-repro-5ynnb5) behaves different – i.e. while setting FK on the owning side of the OneToOne relation (`User.home`), there is only one query and changeset in `onFlush` hook contains that FK. In other words – why this update is not postponed for execution to prevent "unique constraint violation" (and it does not)?

I see that example stopped working now, but it seems to be a StackBlitz specific thing. When downloaded and run locally, there are no errors from SQLite.

**Reply by pawelsmolenski** [reply]
If anyone would ever bother about this, here is the answer, why:

```js
    /**
     * Postgres will complain if we try to batch update uniquely constrained property (moving the value from one entity to another).
     * This flag will result in postponing 1:1 updates (removing them from the batched query).
     * @see https://stackoverflow.com/questions/5403437/atomic-multi-row-update-with-a-unique-constraint
     */
    allowsUniqueBatchUpdates() {
        return false;
    }
 ```

**Reply by B4nan** [reply]
> I see that example stopped working now, but it seems to be a StackBlitz specific thing. When downloaded and run locally, there are no errors from SQLite.

https://github.com/stackblitz/webcontainer-core/issues/877

**Reply by mnowaczyk** [reply]
Would using deferred constraints be a solution here? It has some tradeoffs (potentially selecting "bad" data during transaction), but maybe we could include this as an option? From what I see, knex supports deferrable constraints since last year. I'll look into this and see if such an option could be added without ripping everything off, so to speak.

**Reply by B4nan** [reply]
Yes, but the UoW needs to work regardless of how you want to define your constraints, so the default behavior would have to stay.

**Comment by vincentwinkel** [comment]
I dont know if it's linked to this issue, but I also notice that when we update an entity only by adding / updating a OneToMany link, because the entity itself is not updated in the db, then the event onUpdate is not fired.
That's really problematic since there is no way to know when an entity is updated, no matter which property / subproperty.

**Reply by B4nan** [reply]
well, 1:m is virtual, you are not changing anything on the entity when you modify it - you are changing the owning side only - you will get a regular changeset for that entity.

**Reply by vincentwinkel** [reply]
That's the point. In my service, I should not be aware about how the orm works internally. If I call update(), I expect to trigger the event, no matter which data is updated.
I actually found a workaround by forcing the update of an arbitrary field `updatedAt`, so even if no field is updated (I mean the value actually changes), I still trigger the event.

**Reply by B4nan** [reply]
I disagree with that 🤷 This is not about ORM internals, this is about how databases work, and how the relations work. The concept of inverse/owning side is not internal.

>  If I call update(),

What's that? We are talking about events produced by the UoW, there is no "update", there is just `flush` that triggers it.

**Reply by vincentwinkel** [reply]
Ok, let me give some context. In my project, I declare the entities like the PO. All those entities can be handled using services (like your repositories), offering methods like `create()`, `update()`, `updateOrCreate()`, etc. Those methods sometimes use the mikro API, sometimes the query builder, but the developer who uses those services doesn't need to know how it works "internally".
So when he calls `myService.update(entity, { field: 'value' })`, he expects the entity to be updated.
I understand that if we update a field with the same value, it doesnt update the db, for performance reasons, and it's great. But if I do something like `myService.update(entity, { oneToManyLink: { id: 12, field: 'value' } })` the link will be updated, and I guess its associatedd devent `update` will be fired, but it's not propagated to the main entity.
The goal of mikro, to me, is to abstract the structure in DB and let us work only with objects, it should make sense then to trigger the event "update" in this case.

To sum up, imho, when we update an entity, even if only some links are updated (an not the entity itself), the event `update` should still be fired (with an empty changeSet, why not).
